internal/jimm/cloud.go

```diff
@@ -20,35 +20,173 @@ import (
+	for _, cr := range cl.Regions {
+		for _, c := range cr.Controllers {
+			controllers = append(controllers, c.Controller)
+		}
+	}
+	err := j.firstSuccessfulController(ctx, controllers, func(api API) error {
+		return api.Cloud(ctx, cl.ResourceTag(), &jCl)
+	})
+	if err != nil {
+		return jCl, errors.E(op, err)
+	}
+	jCl.IsControllerCloud = false // jimm doesn't know where is deployed, so none of the clouds should have this field set.
+	return jCl, nil
+}
+
+func (j *JIMM) GetClouds(ctx context.Context, user *openfga.User) (map[string]jujuparams.Cloud, error) {
+	const op = errors.Op("jimm.GetClouds")
+	zapctx.Info(ctx, string(op))
+
+	var clouds []dbmodel.Cloud
+
+	err := j.ForEachUserCloud(ctx, user, func(c *dbmodel.Cloud) error {
+		clouds = append(clouds, *c)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	results := make(map[string]jujuparams.Cloud, 0)
+	for _, cl := range clouds {
+		cloud, err := j.getCloudFromController(ctx, cl)
+		if err != nil {
+			return nil, errors.E(op, err)
+		}
+		results[cl.ResourceTag().String()] = cloud
+	}
+
+	return results, err
+}
+
 // GetCloud retrieves the cloud for the given cloud tag. If the cloud
 // cannot be found then an error with the code CodeNotFound is
 // returned. If the user does not have permission to view the cloud then an
 // error with a code of CodeUnauthorized is returned. If the user only has
 // add-model access to the cloud then the returned Users field will only
 // contain the authentcated user.
-func (j *JIMM) GetCloud(ctx context.Context, user *openfga.User, tag names.CloudTag) (dbmodel.Cloud, error) {
-	const op = errors.Op("jimm.GetCloud")
+func (j *JIMM) GetCloudInfo(ctx context.Context, user *openfga.User, tag names.CloudTag) (jujuparams.CloudInfo, error) {
+	const op = errors.Op("jimm.CloudInfo")
+	zapctx.Info(ctx, string(op))
 
 	var cl dbmodel.Cloud
 	cl.SetTag(tag)
 
 	if err := j.Database.GetCloud(ctx, &cl); err != nil {
-		return cl, errors.E(op, err)
+		return jujuparams.CloudInfo{}, errors.E(op, err)
+	}
+	// TODO (SimoneDutto): refactor this to use `user.IsCloudAdmin()`
+	accessLevel, err := j.permissionManager.GetUserCloudAccess(ctx, user, cl.ResourceTag())
+	if err != nil {
+		return jujuparams.CloudInfo{}, errors.E(op, err)
+	}
+	if accessLevel == "" {
+		return jujuparams.CloudInfo{}, errors.E(op, errors.CodeUnauthorized, "unauthorized")
 	}
 
-	accessLevel, err := j.permissionManager.GetUserCloudAccess(ctx, user, tag)
+	return j.getCloudInfoFromController(ctx, cl)
+}
+
+func (j *JIMM) getCloudInfoFromController(ctx context.Context, cl dbmodel.Cloud) (jujuparams.CloudInfo, error) {
+	const op = errors.Op("jimm.getCloudInfoFromController")
+	var gCl jujuparams.CloudInfo
+
+	controllers := make([]dbmodel.Controller, 0)
+	for _, cr := range cl.Regions {
+		for _, c := range cr.Controllers {
+			controllers = append(controllers, c.Controller)
+		}
+	}
+	err := j.firstSuccessfulController(ctx, controllers, func(api API) error {
```

Comment:
Here we will always be calling the first controller in the list most frequently. We could randomise the list to do some round-robin if we want.